packages/api-request-builder/src/default-params.js

```diff
@@ -62,23 +63,104 @@ export function getDefaultSearchParams (): ServiceBuilderDefaultParams {
+
+/**
+ * Set the supplied parameters given the current service object.
+ *
+ * @return {void}
+ */
+export function setParams (params: ServiceBuilderParams) {
+  // verify params
+  const knownKeys = [
+    'expand', 'id', 'key', 'customerId', 'sort', 'page', 'perPage', 'staged',
+    'priceCurrency', 'priceCountry', 'priceCustomerGroup', 'priceChannel',
+    'text', 'fuzzy', 'fuzzyLevel', 'markMatchingVariants', 'facet', 'filter',
+    'filterByQuery', 'filterByFacets', 'searchKeywords', 'where', 'version',
+  ]
+  Object.keys(params).forEach((key: string) => {
+    if (!knownKeys.includes(key)) throw new Error(`Unknown key "${key}"`)
+  })
+
+  // query-expand
+  if (params.expand)
+    params.expand.forEach((expansion: string) => { this.expand(expansion) })
+
+  // query-id
+  if (hasKey(params, 'id')) this.byId(params.id)
+  if (hasKey(params, 'key')) this.byKey(params.key)
+  if (hasKey(params, 'customerId')) this.byCustomerId(params.customerId)
+
+  // query-page
+  if (params.sort) params.sort.forEach(
+    (sortDesc: { by: string, direction: 'asc' | 'desc' }) => {
+      this.sort(sortDesc.by, sortDesc.direction === 'asc')
+    },
+  )
+  if (hasKey(params, 'page')) this.page(params.page)
+  if (hasKey(params, 'perPage')) this.perPage(params.perPage)
+
+  // query-projection
+  if (hasKey(params, 'staged')) this.staged(params.staged)
+  if (hasKey(params, 'priceCurrency')) this.priceCurrency(params.priceCurrency)
+  if (hasKey(params, 'priceCountry')) this.priceCountry(params.priceCountry)
+  if (hasKey(params, 'priceCustomerGroup'))
+    this.priceCustomerGroup(params.priceCustomerGroup)
+  if (hasKey(params, 'priceChannel')) this.priceChannel(params.priceChannel)
+
+  // query-search
+  if (params.text) this.text(params.text.value, params.text.language)
+  if (params.fuzzy) this.fuzzy() // boolean switch
+  if (hasKey(params, 'fuzzyLevel')) this.fuzzyLevel(params.fuzzyLevel)
+  if (params.markMatchingVariants) this.markMatchingVariants() // boolean switch
+  if (params.facet)
+    params.facet.forEach((facet: string) => { this.facet(facet) })
+  if (params.filter)
+    params.filter.forEach((filter: string) => { this.filter(filter) })
+  if (params.filterByQuery)
+    params.filterByQuery.forEach(
+      (query: string) => { this.filterByQuery(query) },
+    )
+  if (params.filterByFacets)
+    params.filterByFacets.forEach(
+      (facet: string) => { this.filterByFacets(facet) },
+    )
+
+  // query-suggest
+  if (params.searchKeywords)
+    params.searchKeywords.forEach(
+      (searchKeyword: {value: string, language: string }) => {
+        this.searchKeywords(searchKeyword.value, searchKeyword.language)
+      },
+    )
+
+  // query
+  if (params.where)
+    params.where.forEach((predicate: string) => { this.where(predicate) })
+  if (hasKey(params, 'whereOperator')) this.whereOperator(params.whereOperator)
+
+  // version
+  if (hasKey(params, 'version')) this.withVersion(params.version)
+}
```

Comment:
<details>
<summary>I also considered doing this approach (expand me)</summary>

```js
/**
 * Set the supplied parameters given the current service object.
 *
 * @return {void}
 */
export function setParams (params: ServiceBuilderDefaultParams) {
  Object.entries(params).forEach(([paramKey, paramValue]: [string, mixed]) => {
    switch (paramKey) {
      // query-expand
      case 'expand':
        if (!Array.isArray(paramValue)) throw new Error('Expected array')
        paramValue.forEach((expansion: string) => { this.expand(expansion) })
        break

      // query-id
      case 'id':
        this.byId(paramValue)
        break
      case 'key':
        this.byKey(paramValue)
        break
      case 'customerId':
        this.byCustomerId(paramValue)
        break

      // query-page
      case 'sort':
        paramValue.forEach(
          (sortDesc: { by: string, direction: 'asc' | 'desc' }) => {
            this.sort(sortDesc.by, sortDesc.direction === 'asc')
          },
        )
        break
      case 'page':
        this.page(paramValue)
        break
      case 'perPage':
        this.perPage(paramValue)
        break

      // query-projection
      case 'staged':
        this.staged(paramValue)
        break
      case 'priceCurrency':
        this.priceCurrency(paramValue)
        break
      case 'priceCountry':
        this.priceCountry(paramValue)
        break
      case 'priceCustomerGroup':
        this.priceCustomerGroup(paramValue)
        break
      case 'priceChannel':
        this.priceChannel(paramValue)
        break

      // query-search
      case 'text':
        this.text(paramValue.value, paramValue.language)
        break
      case 'fuzzy':
        if (paramValue) this.fuzzy()
        break
      case 'fuzzyLevel':
        this.fuzzyLevel(paramValue)
        break
      case 'markMatchingVariants':
        if (paramValue) this.markMatchingVariants()
        break
      case 'facet':
        paramValue.forEach((facet: string) => { this.facet(facet) })
        break
      case 'filter':
        paramValue.forEach((filter: string) => { this.filter(filter) })
        break
      case 'filterByQuery':
        paramValue.forEach((query: string) => { this.filterByQuery(query) })
        break
      case 'filterByFacets':
        paramValue.forEach((facet: string) => { this.filterByFacets(facet) })
        break

      // query-suggest
      case 'searchKeywords':
        paramValue.forEach(
          (searchKeyword: {value: string, language: string }) => {
            this.searchKeywords(searchKeyword.value, searchKeyword.language)
          },
        )
        break

      // query
      case 'where':
        paramValue.forEach((predicate: string) => { this.where(predicate) })
        break
      case 'whereOperator':
        this.whereOperator(paramValue)
        break

      // version
      case 'version':
        this.withVersion(paramValue)
        break

      default: throw new Error(`Unknown key "${paramKey}"`)
    }
  })
}
```

</details>

--- 


That `switch` approach
- makes the `knownKeys` array unnecessary
- makes it hard in case we want to have nested keys (we'd need nested `switch`-statements)

Not too sure which is better. Both work.

Reply:
Thanks a lot for this PR (loving it). I think the `switch` approach will be better for this and is in my opinion, more concise, and also favours the flat key approach (I think page, `perPage` and `sort` should not be nested)

Reply:
I forgot one downside of the `switch` approach:
- we loose control over the order in which the setters are called as the calls will happen in the order of the keys of the object. `Object.entries` traverses the objects in an arbitrary order, so the application using the library doesn't gain control either. But we probably don't need the control anyways. In case we do, we could always execute some before/after the `switch` statement.

Reply:
True. And the besides the `id`, `key` and `customerId`, the order in which the setters are called does not really matter. But then, a user should not pass in parameters having more than one of these properties.

Reply:
At least for now. We can go with the `if` statements, in case future developments make the order of execution necessary